store/postgres/src/deployment_store.rs

```diff
@@ -357,52 +357,111 @@ impl DeploymentStore {
-                            e
-                        )
-                        .into()
-                    }),
-            }?;
-            count += n;
+            }
+        }
+        // apply modification groups
+        for (entity_type, entities) in inserts.iter_mut() {
+            count += self.insert_entities(entity_type, entities, conn, layout, ptr, &stopwatch)?;
+        }
+        for (entity_type, entities) in overwrites.iter_mut() {
+            count +=
+                self.overwrite_entities(entity_type, entities, conn, layout, ptr, &stopwatch)?;
+        }
+        for (entity_type, entity_keys) in removals.into_iter() {
+            count +=
+                self.remove_entities(entity_type, entity_keys, conn, layout, ptr, &stopwatch)?;
         }
         Ok(count)
     }
 
+    fn insert_entities(
+        &self,
+        entity_type: &EntityType,
+        data: &mut Vec<(EntityKey, Entity)>,
+        conn: &PgConnection,
+        layout: &Layout,
+        ptr: &EthereumBlockPointer,
+        stopwatch: &StopwatchMetrics,
+    ) -> Result<i32, StoreError> {
+        let section = stopwatch.start_section("check_interface_entity_uniqueness");
+        for (key, _) in data.iter() {
+            // WARNING: This will potentially execute 2 queries for each entity key.
+            self.check_interface_entity_uniqueness(conn, layout, key)?;
+        }
+        section.end();
+
+        let _section = stopwatch.start_section("apply_entity_modifications_insert");
+        layout
+            .insert(conn, entity_type, data, block_number(ptr))
+            .map(|_| 1)
+    }
+
+    fn overwrite_entities(
+        &self,
+        entity_type: &EntityType,
+        data: &mut Vec<(EntityKey, Entity)>,
+        conn: &PgConnection,
+        layout: &Layout,
+        ptr: &EthereumBlockPointer,
+        stopwatch: &StopwatchMetrics,
+    ) -> Result<i32, StoreError> {
+        let section = stopwatch.start_section("check_interface_entity_uniqueness");
+        for (key, _) in data.iter() {
+            // WARNING: This will potentially execute 2 queries for each entity key.
+            self.check_interface_entity_uniqueness(conn, layout, key)?;
+        }
+        section.end();
+
+        let _section = stopwatch.start_section("apply_entity_modifications_update");
+        layout
+            .update(conn, entity_type, data, block_number(ptr))
+            .map(|_| 0)
+    }
+
+    fn remove_entities(
+        &self,
+        entity_type: EntityType,
+        entity_keys: Vec<EntityKey>,
+        conn: &PgConnection,
+        layout: &Layout,
+        ptr: &EthereumBlockPointer,
+        stopwatch: &StopwatchMetrics,
+    ) -> Result<i32, StoreError> {
+        let _section = stopwatch.start_section("apply_entity_modifications_delete");
+        layout
+            .delete(conn, entity_type, &entity_keys, block_number(ptr))
+            // This conversion is ok since n will only be 0 or 1
+            .map(|n| -(n as i32))
```

Comment:
Similarly, this needs to return the negative of the number of entities deleted to make sure the overall entity count is correct.